src/data/shortlist.cpp

```diff
@@ -0,0 +1,396 @@
+  // This is necessary until intgemm supports non-multiple-of-eight matrices.
+  for (size_t i = firstNum_; i < trgVocabSize && trgTruthTableOnes%8!=0; i++){
+    if (!trgTruthTable[i]){
+      trgTruthTable[i] = 1;
+      trgTruthTableOnes++;
+    }
+  }
+
+  // turn selected indices into vector and sort (Bucket sort: O(V))
+  std::vector<WordIndex> indices;
+  for (WordIndex i = 0; i < trgVocabSize; i++) {
+    if(trgTruthTable[i])
+      indices.push_back(i);
+  }
+
+  return New<Shortlist>(indices);
+}
+
+void BinaryShortlistGenerator::dump(const std::string& fileName) const {
+  ABORT_IF(mmapMem_.is_open(),"No need to dump again");
+  LOG(info, "[data] Saving binary shortlist dump to {}", fileName);
+  std::vector<char> blob = generateBlob();
+  saveBlobToFile(blob,fileName);
+}
+
+void BinaryShortlistGenerator::import(const std::string& filename, double threshold) {
+  io::InputFileStream in(filename);
+  std::string src, trg;
+
+  // Read text file
+  std::vector<std::unordered_map<WordIndex, float>> srcTgtProbTable;
+  float prob;
+
+  while(in >> trg >> src >> prob) {
+    if(src == "NULL" || trg == "NULL")
+      continue;
+
+    auto sId = (*srcVocab_)[src].toWordIndex();
+    auto tId = (*trgVocab_)[trg].toWordIndex();
+    if(srcTgtProbTable.size() <= sId)
+      srcTgtProbTable.resize(sId + 1);
+    if(srcTgtProbTable[sId][tId] < prob)
+      srcTgtProbTable[sId][tId] = prob;
+  }
+
+  // Create priority queue and count
+  std::vector<std::priority_queue<std::pair<double, WordIndex>>> vpq;
+  uint64_t shortListSize = 0;
+
+  vpq.resize(srcTgtProbTable.size());
+  for(WordIndex sId = 0; sId < srcTgtProbTable.size(); sId++) {
+    uint64_t shortListSizeCurrent = 0;
+    for(auto entry : srcTgtProbTable[sId]) {
+      if (entry.first>=threshold) {
+        vpq[sId].push(std::make_pair(entry.second, entry.first));
+        if(shortListSizeCurrent < bestNum_)
+          shortListSizeCurrent++;
+      }
+    }
+    shortListSize += shortListSizeCurrent;
+  }
+
+  uint64_t wordToOffsetSize = vpq.size()+1;
+  uint64_t* wordToOffset = new uint64_t[wordToOffsetSize];
+  WordIndex* shortLists = new WordIndex[shortListSize];
+  WordIndex shortlistIdx = 0;
+  for(size_t i=0; i< wordToOffsetSize -1; i++) {
+    wordToOffset[i] = shortlistIdx;
+    for(int popcnt = 0; popcnt < bestNum_ && !vpq[i].empty(); popcnt++) {
+      shortLists[shortlistIdx] = vpq[i].top().second;
+      shortlistIdx++;
+      vpq[i].pop();
+    }
+  }
+  wordToOffset[wordToOffsetSize-1] = shortlistIdx;
+
+  // Sort word indices for each shortlist
+  for(int i = 1; i < wordToOffsetSize; i++) {
+    std::sort(&shortLists[wordToOffset[i-1]], &shortLists[wordToOffset[i]]);
+  }
+
+  wordToOffset_ = wordToOffset;
+  shortLists_ = shortLists;
+  wordToOffsetSize_ = wordToOffsetSize;
+  shortListsSize_ = shortListSize;
+}
+
+std::vector<char> BinaryShortlistGenerator::generateBlob() const{
+  // Build the body
+  std::vector<char> body;
```

Comment:
You should know exactly how many bytes will be needed so can size this properly.  In fact the file could be truncated to length, mmaped with sharing, and written to.

Reply:
I do not understand. Shall I specify a size to the blob when it is created?

Reply:
If you tell the blob the size upfront, it won't waste time reallocating memory.